### PR TITLE
Make sure flash resets on subsequent requests

### LINKED ISSUE #404: Make flash messages rotate again
Flash messages don't seem to reset automatically anymore.
- [x] Make sure flash cookie is reset after request (rotated)
- [x] Add tests with two empty requests to make sure it is reset/emptied
- [x] Remove `Flash.set_alert None` in flash test, should be covered by the automatic rotation

### PR DESCRIPTION
**Pull request by jerben**
Fix #404

### CODE REVIEW

File: sihl/test/web_flash.ml [hunk 1]
```diff
@@ -80,44 +93,51 @@ let remove_flash_message _ () =
...
         let custom = Sihl.Web.Flash.find_custom req in
-        Alcotest.(check (option string) "has no alert" None alert);
+        Alcotest.(check (option string) "has alert" (Some "nextfoo") alert);
         Alcotest.(check (option string) "has custom" (Some "hello") custom);
         let res = Opium.Response.of_plain_text "" in
         Lwt.return res)
       req
   in
-  let req = Opium.Request.get "" in
+  (* Simulating the browser sending the last Set-Cookie *)
```
Comment (aronerben) [inline sihl/test/web_flash.ml:131]: What was the comment scheme we agreed on? Imperative and passive? If yes, please rewrite to something like:
`Simulate the browser sending the last Set-Cookie`

**Review by aronerben — CHANGES_REQUESTED** [review]
Good implementation, good tests, thanks

File: sihl/src/web_flash.ml [hunk 2]
```diff
@@ -11,7 +11,19 @@ module Flash = struct
 
   let empty = { alert = None; notice = None; custom = None }
 
-  let of_yojson json =
+  let is_empty (flash : t) : bool =
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/web_flash.ml:14]: I add them for private functions for better type error messages.

File: sihl/src/web_flash.ml [hunk 3]
```diff
@@ -123,31 +135,32 @@ let decode_flash cookie_key req =
...
 ;;
 
-let persist_flash cookie_key resp =
+let persist_flash old_flash cookie_key resp =
   let flash = Opium.Context.find Env.key resp.Opium.Response.env in
   match flash with
-  | None -> (* No need to touch the flash cookie *) resp
+  | None ->
+    if Flash.is_empty old_flash
+    then (* We didn't touch the flash, no need to set cookie *)
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/web_flash.ml:173]: We can use passive, I don't think we talked about it yet. Fixed.

File: sihl/src/web_flash.ml [hunk 4]
```diff
@@ -123,31 +135,32 @@ let decode_flash cookie_key req =
...
+      resp
+    else (
+      (* Clear flash *)
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/web_flash.ml:176]: Bad comment, the flash cookie needs to be cleared. Fixed.

File: sihl/src/web_flash.ml [hunk 5]
```diff
@@ -123,31 +135,32 @@ let decode_flash cookie_key req =
...
+      let resp = Opium.Response.add_cookie_or_replace cookie resp in
+      resp)
   | Some flash ->
-    (* The flash changed, we need to persist the new flash in the cookie *)
-    let cookie_value = Flash.to_json flash in
-    let cookie = cookie_key, cookie_value in
-    let resp = Opium.Response.add_cookie_or_replace cookie resp in
-    resp
+    if Flash.equals old_flash flash
+    then (* We didn't touch the flash, no need to set cookie *)
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/src/web_flash.ml:182]: We can use passive, I don't think we talked about it yet. Fixed.

File: sihl/test/web_flash.ml [hunk 6]
```diff
@@ -1,22 +1,46 @@
...
+      (fun _ ->
+        let res =
+          Opium.Response.of_plain_text ""
+          |> Sihl.Web.Flash.set_alert (Some "foobar")
+        in
         Lwt.return res)
       req
   in
-  (* we need to simulate the browser sending back the cookie *)
+  (* Simulating the browser sending the last Set-Cookie *)
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/test/web_flash.ml:43]: Fixed.

File: sihl/test/web_flash.ml [hunk 7]
```diff
@@ -26,47 +50,36 @@ let set_and_read_flash_message _ () =
...
         Alcotest.(check (option string) "has alert" (Some "foobar") alert);
-        Alcotest.(check (option string) "has no notice" None notice);
-        let res = Opium.Response.of_plain_text "" in
-        let res = Sihl.Web.Flash.set_alert (Some "nextfoo") res in
-        let res = Sihl.Web.Flash.set_custom (Some "hello") res in
-        Lwt.return res)
+        Lwt.return @@ Opium.Response.of_plain_text "")
       req
   in
+  (* Simulating the browser sending the last Set-Cookie *)
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/test/web_flash.ml:57]: Fixed.

File: sihl/test/web_flash.ml [hunk 8]
```diff
@@ -26,47 +50,36 @@ let set_and_read_flash_message _ () =
...
-        Lwt.return res)
+        Alcotest.(check (option string) "alert was cleared" None alert);
+        Lwt.return @@ Opium.Response.of_plain_text "")
       req
   in
+  Alcotest.(
+    check
+      (pair string string)
+      "flash is empty"
+      (* default empty flash with default test secret *)
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/test/web_flash.ml:75]: Yessir, fixed.

File: sihl/test/web_flash.ml [hunk 9]
```diff
@@ -80,44 +93,51 @@ let remove_flash_message _ () =
         Lwt.return res)
       req
   in
-  (* we need to simulate the browser sending back the cookie *)
+  (* Simulating the browser sending the last Set-Cookie *)
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/test/web_flash.ml:96]: Fixed.

File: sihl/test/web_flash.ml [hunk 10]
```diff
@@ -80,44 +93,51 @@ let remove_flash_message _ () =
...
+        Alcotest.(check (option string) "has alert" (Some "foobar") alert);
+        Alcotest.(check (option string) "has no notice" None notice);
         let res = Opium.Response.of_plain_text "" in
-        let res = Sihl.Web.Flash.set_alert None res in
+        let res = Sihl.Web.Flash.set_alert (Some "nextfoo") res in
         let res = Sihl.Web.Flash.set_custom (Some "hello") res in
         Lwt.return res)
       req
   in
+  (* Simulating the browser sending the last Set-Cookie *)
```
[the remark this answers is not in the document]
Comment (jerben) [reply sihl/test/web_flash.ml:115]: Fixed.

File: sihl/test/web_flash.ml (hunk 1 above)
Comment (jerben) [reply sihl/test/web_flash.ml:131]: Fixed.

File: sihl/src/web_flash.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (aronerben) [reply sihl/src/web_flash.ml]: Ok, let's follow this format, if you agree.

**Review by aronerben — APPROVED** [review]
